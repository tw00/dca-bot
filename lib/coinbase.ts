/* eslint-disable @typescript-eslint/ban-ts-comment */
import WebSocket from "ws";

export interface ICoinbaseTickerData {
  sequence: number;
  product_id: string;
  price: string;
  open_24h: string;
  volume_24h: string;
  low_24h: string;
  high_24h: string;
  volume_30d: string;
  best_bid: string;
  best_ask: string;
  side: string;
  time: string;
  trade_id: number;
  last_size: string;
}

type SubscriptionCallback = (data: ICoinbaseTickerData) => void;

export default class CoinbaseWebsocketClient {
  ws: WebSocket;
  subscriber: {
    [key: string]: SubscriptionCallback;
  };

  // @ts-ignore
  constructor(): Promise<CoinbaseWebsocketClient> {
    const url = "wss://ws-feed.pro.coinbase.com";
    const ws = new WebSocket(url);

    this.ws = ws;
    this.subscriber = {};
    const ready = new Promise((resolve) => {
      ws.on("open", () => {
        resolve(this);
      });
    });

    ws.on("message", (data) => {
      this.onMessage(JSON.parse(data) as ICoinbaseTickerData);
    });

    // @ts-ignore
    return ready;
  }

  send(data: unknown): void {
    this.ws.send(JSON.stringify(data));
  }

  subscribe(ticker: string, callback: SubscriptionCallback): void {
    this.subscriber[ticker] = callback;
    this.send({
      type: "subscribe",
      product_ids: [ticker],
      channels: [
        // "level2",
        // "heartbeat",
        {
          name: "ticker",
          product_ids: [ticker],
        },
      ],
    });
  }

  onMessage(data: ICoinbaseTickerData): void {
    const cb = this.subscriber[data.product_id];
    if (cb) {
      cb(data);
    }
  }
}
