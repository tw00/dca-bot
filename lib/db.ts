import { default as levelup, LevelUp } from "levelup";
import { default as leveldown } from "leveldown";

export interface TimeIndexableData {
  type: string;
  key?: number;
  time: number;
}

export interface CoinbaseTickerData extends TimeIndexableData {
  type: "ticker";
  sequence: number;
  product_id: string;
  price: string; // TODO: convert to number
  open_24h: string;
  volume_24h: string;
  low_24h: string;
  high_24h: string;
  volume_30d: string;
  best_bid: string;
  best_ask: string;
  side: string;
  // time: Date; // TODO !!
  trade_id: number;
  last_size: string;
}

export interface CoinbaseHistoricalData extends TimeIndexableData {
  type: "candle";
  low: number;
  high: number;
  open: number;
  close: number;
  volume: number;
}

export enum DBType {
  HISTORIC = "historic",
  TICK = "tick",
}

type DataCallback<T> = (data: T) => void;

export default class DB<T extends CoinbaseHistoricalData | CoinbaseTickerData> {
  db: LevelUp;
  type: DBType;

  constructor(name, type: DBType) {
    const db = levelup(leveldown(`./db/${name}-${type}`));
    this.db = db;
    this.type = type;
  }

  append(data: T) {
    const key = +new Date(data.time);
    const value = JSON.stringify(data);
    this.db.put(key, value);
  }

  delete(key: number) {
    this.db.del(key);
  }

  readStream(dataCb: DataCallback<T>, endCb) {
    this.db
      .createReadStream()
      .on("data", (raw) => {
        const data = JSON.parse(raw.value.toString()) as T;
        data.key = raw.key.toString();
        dataCb(data);
      })
      .on("error", (err) => {
        throw err;
      })
      .on("end", () => {
        endCb();
      });
  }

  read(from: number = null, to: number = null): Promise<T[]> {
    const result = [];
    const options = {
      ...(from ? { gte: from } : {}),
      ...(to ? { lte: to } : {}),
    };

    return new Promise((resolve, reject) => {
      this.db
        .createReadStream(options)
        .on("data", (raw) => {
          const data = JSON.parse(raw.value.toString()) as T;
          data.key = raw.key.toString();
          result.push(data);
        })
        .on("error", (err) => {
          reject(err);
        })
        .on("end", () => {
          resolve(result);
        });
    });
  }
}
