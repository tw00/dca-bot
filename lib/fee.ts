import { strict as assert } from "assert";

interface IFeeTabEntry {
  thres: number;
  tier: string;
  takerFee: number; // percentage
  makerFee: number; // percentage
}

const feeCoinbase: IFeeTabEntry[] = [
  { thres: 0, tier: "Up to $10k", takerFee: 0.5, makerFee: 0.5 },
  { thres: 10 * 1e3, tier: "$10k - $50k", takerFee: 0.35, makerFee: 0.35 },
  { thres: 50 * 1e3, tier: "$50k - $100k", takerFee: 0.25, makerFee: 0.15 },
  { thres: 100 * 1e3, tier: "$100k - $1m", takerFee: 0.2, makerFee: 0.1 },
  { thres: 1 * 1e6, tier: "$1m - $10m", takerFee: 0.18, makerFee: 0.08 },
  { thres: 10 * 1e6, tier: "$10m - $50m", takerFee: 0.15, makerFee: 0.05 },
  { thres: 50 * 1e6, tier: "$50m - $100m", takerFee: 0.1, makerFee: 0.0 },
  { thres: 100 * 1e6, tier: "$100m - $300m", takerFee: 0.07, makerFee: 0.0 },
  { thres: 300 * 1e6, tier: "$300m - $500m", takerFee: 0.05, makerFee: 0.0 },
  { thres: 500 * 1e6, tier: "$500m - $1b", takerFee: 0.04, makerFee: 0.0 },
  { thres: 1e9, tier: "$1b+", takerFee: 0.04, makerFee: 0.0 },
];

const fees = {
  coinbase: feeCoinbase,
};

export function getDiscountedFee(
  volumeLast30Days: number,
  exchange = "coinbase"
): IFeeTabEntry {
  assert(volumeLast30Days >= 0, "volumeLast30Days must be positive");
  const fee = fees[exchange];
  const idx = fee.findIndex((x) => volumeLast30Days < x.thres);
  return fee[(idx === -1 ? fee.length : idx) - 1];
}
