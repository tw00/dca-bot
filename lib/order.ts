export enum OrderType {
  STOP = "stop",
  MARKET = "market",
  // TRAILING_STOP,
}

export enum OrderSide {
  BUY = "buy",
  SELL = "sell",
}

export default class Order {
  side: OrderSide;
  type: OrderType;
  symbol: string;
  price: number;
  amount: number;

  constructor(
    side: OrderSide,
    symbol: string,
    price: number,
    amount: number,
    type: OrderType = OrderType.STOP
  ) {
    this.side = side;
    this.symbol = symbol;
    this.price = price;
    this.amount = amount;
    this.type = type;
  }
}
