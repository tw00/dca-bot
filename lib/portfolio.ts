export interface PositionUpdate {
  symbol: string;
  amount: number;
  price: number;
  time?: Date;
}

export interface Transaction extends PositionUpdate {
  balance?: number;
}

interface PositionMap {
  [key: string]: number;
}

export default class Portfolio {
  positions: PositionMap;
  transactions: Transaction[];

  constructor() {
    this.positions = {};
    this.transactions = [];
  }

  addPosition(position: PositionUpdate): void {
    // console.log("New position", position);

    if (!(position.symbol in this.positions)) {
      this.positions[position.symbol] = 0;
    }
    this.positions[position.symbol] += position.amount;
    this.transactions.push({
      ...position,
      balance: this.getFunds(position.symbol),
    });
  }

  getFunds(symbol) {
    return this.positions[symbol] || 0;
  }

  print() {
    console.log("=".repeat(40));
    console.log("Transactions:", this.transactions);
    console.log("Positions:", this.positions);
    console.log("=".repeat(40));
  }
}
