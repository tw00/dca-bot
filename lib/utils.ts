export function guessPrice(candle) {
  return (candle.low + candle.high) / 2;
}

export function range(from, to) {
  return [...Array(to - from + 1).keys()].map((v) => v + from);
}
