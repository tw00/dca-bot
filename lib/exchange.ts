import Order, { OrderSide, OrderType } from "./order";
import Portfolio from "./portfolio";
import { getDiscountedFee } from "../lib/fee";

export interface TickInfo {
  symbol: string;
  time: number;
  price: number;
}

export interface IExchangeOptions {
  fee?: number;
  volumeLast30Days?: number;
}

export default class Exchange {
  orderbook: Order[];
  executedOrder: Order[];
  lastTick: TickInfo | null;
  portfolio: Portfolio;

  constructor({ volumeLast30Days, fee }: IExchangeOptions) {
    this.orderbook = [];
    this.executedOrder = [];
    this.lastTick = null;
    this.portfolio = new Portfolio({
      fee: volumeLast30Days ? getDiscountedFee(volumeLast30Days).makerFee : fee,
    });
  }

  addOrder(order) {
    this.orderbook.push(order);
  }

  clearOrder(order) {
    this.executedOrder.push(order);
    this.orderbook = this.orderbook.filter((o) => o !== order);
  }

  closePosition(symbol) {
    if (symbol == this.lastTick.symbol) {
      const amount = -this.portfolio.getFunds(symbol);
      this.portfolio.transaction(symbol, "USD", this.lastTick.price, amount, {
        includeFee: false,
      });
      // TODO: use market order?
    }
  }

  feed(data: TickInfo): void {
    // TODO: Slipage

    this.lastTick = data;

    this.orderbook.forEach((order) => {
      const transactQuote = (amount: number) => {
        this.portfolio.transaction(order.symbol, "USD", data.price, amount, {
          includeFee: true,
          time: new Date(data.time * 1000),
        });
        this.clearOrder(order);
      };

      if (order.sellAll) {
        order.amount = this.portfolio.getFunds(order.symbol);
      }

      if (order.symbol !== data.symbol) {
        return;
      }

      if (order.type === OrderType.MARKET) {
        if (order.side === OrderSide.BUY) {
          transactQuote(order.amount);
        }
        if (order.side === OrderSide.SELL) {
          transactQuote(-order.amount);
        }
      }

      if (order.type === OrderType.STOP) {
        if (order.side === OrderSide.BUY) {
          if (order.price >= data.price) {
            transactQuote(order.amount);
          }
        }

        if (order.side === OrderSide.SELL) {
          if (order.price <= data.price) {
            transactQuote(-order.amount);
          }
        }
      }
    });
  }
}
