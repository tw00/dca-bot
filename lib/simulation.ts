import Exchange, { TickInfo } from "./exchange";
import DB, { DBType, ITickerData, IHistoricalData } from "./db";
import Bot from "../bots/bot";
import { guessPrice } from "./utils";

interface ISimulationOptions {
  from?: Date;
  to?: Date;
  type?: DBType;
  symbol: string;
  fee: number;
}

const SHOW_TICKS = false;

export default class Simulation {
  exchange: Exchange;
  bots: Bot[];
  options: ISimulationOptions;

  constructor(options: ISimulationOptions) {
    this.exchange = new Exchange({ fee: options.fee });
    this.bots = [];
    this.options = {
      type: DBType.TICK,
      symbol: "ETH",
      ...options,
    };
  }

  addBot(bot: Bot): void {
    this.bots.push(bot.withFee(this.options.fee));
  }

  async init(initialCapital: number = 1000): Promise<void> {
    this.exchange.portfolio.updatePosition({
      amount: initialCapital,
      symbol: "USD",
    });
  }

  async run(): Promise<void> {
    return new Promise((resolve) => {
      const db = new DB<ITickerData>(
        `${this.options.symbol}-USD`,
        this.options.type
      );

      const streamOptions = {
        from: +this.options.from,
        to: +this.options.to,
      };

      db.readStream(
        this.tick.bind(this),
        () => {
          this.exchange.closePosition(this.options.symbol);
          resolve();
        },
        streamOptions
      );
    });
  }

  tick(data: ITickerData | IHistoricalData): void {
    const tick: TickInfo =
      this.options.type === DBType.TICK
        ? {
            price: Number((data as ITickerData).price),
            time: +new Date((data as ITickerData).time),
            symbol: this.options.symbol,
          }
        : {
            price: guessPrice(data as IHistoricalData),
            time: (data as IHistoricalData).time,
            symbol: this.options.symbol,
          };

    for (const bot of this.bots) {
      const orders = bot.decide(tick);

      if (orders) {
        console.log("Bot created order", orders);
        orders.forEach((order) => this.exchange.addOrder(order));
      }
    }

    if (SHOW_TICKS) console.log("tick:", tick);
    this.exchange.feed(tick);
  }
}
