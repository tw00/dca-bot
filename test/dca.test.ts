import DCABot from "../bots/dca";
import { OrderSide, OrderType } from "../lib/order";

const roundBy = (num, digits) =>
  Math.round(num * 10 ** digits + Number.EPSILON) / 10 ** digits;

describe("DCA Bot", () => {
  let bot;

  beforeAll(() => {
    bot = new DCABot("TradeAltCoins", { symbol: "BTC" });
  });

  it("initialize correctly", async () => {
    expect(bot.active).toBe(0);
    expect(bot.amountBought).toBe(0);
    expect(bot.amountSpend).toBe(0);
    expect(bot.lastPrice).toBe(0);
    expect(bot.averagePrice).toBe(0);
  });

  it("make base order ASAP", async () => {
    const price = 1000.0;
    const [order] = bot.decide({ symbol: "BTC", time: 0, price });
    expect(bot.active).toBe(1);
    expect(bot.amountSpend).toBe(bot.config.baseOrder);
    expect(bot.amountBought).toBe(bot.config.baseOrder / price);
    expect(bot.averagePrice).toBe(price);
    expect(order.side).toBe(OrderSide.BUY);
    expect(order.type).toBe(OrderType.MARKET);
    expect(order.amount).toBe(0.025);
  });

  it.skip("not buy", async () => {
    bot.decide({ symbol: "BTC", time: 1, price: 1000.0 - 10 });
    expect(bot.active).toBe(1);
  });

  it("buy", async () => {
    const [order] = bot.decide({ symbol: "BTC", time: 2, price: 1000.0 - 25 });
    console.log("order", order);
    // bot.decide({ symbol: "BTC", time: 2, price: 100.0 });
  });

  it("crunch", () => {
    const steps = bot.crunch(1.1998);
    expect(steps).toMatchSnapshot();

    const { maxBotUsageBase, maxDeviation } = bot.summary(1.1998);
    expect(roundBy(maxBotUsageBase, 2)).toBe(365.1);
    expect(roundBy(maxDeviation, 2)).toBe(12.3);
  });
});
