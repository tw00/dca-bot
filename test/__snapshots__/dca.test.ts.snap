// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DCA Bot crunch 1`] = `
Array [
  Object {
    "amountSpendBase": 20.836806134355726,
    "averagePrice": 1.1998,
    "maxDrawdownBase": 25,
    "order": 25,
    "orderNo": 0,
    "price": 1.1998,
    "profitTarget": 1.2177969999999998,
    "safetyOrderDeviation": 0,
  },
  Object {
    "amountSpendBase": 63.36090028610211,
    "averagePrice": 1.1836953020134227,
    "maxDrawdownBase": 75,
    "order": 50,
    "orderNo": 1,
    "price": 1.1758039999999998,
    "profitTarget": 1.2014507315436238,
    "safetyOrderDeviation": 2,
  },
  Object {
    "amountSpendBase": 108.950911624034,
    "averagePrice": 1.170251795964543,
    "maxDrawdownBase": 127.5,
    "order": 52.5,
    "orderNo": 2,
    "price": 1.15156804,
    "profitTarget": 1.1878055729040111,
    "safetyOrderDeviation": 4.02,
  },
  Object {
    "amountSpendBase": 157.86006144791523,
    "averagePrice": 1.1568790631711225,
    "maxDrawdownBase": 182.625,
    "order": 55.125,
    "orderNo": 3,
    "price": 1.1270897203999999,
    "profitTarget": 1.1742322491186892,
    "safetyOrderDeviation": 6.0602,
  },
  Object {
    "amountSpendBase": 210.36641375909574,
    "averagePrice": 1.1432730429840352,
    "maxDrawdownBase": 240.50625,
    "order": 57.88125,
    "orderNo": 4,
    "price": 1.102366617604,
    "profitTarget": 1.1604221386287956,
    "safetyOrderDeviation": 8.120802000000001,
  },
  Object {
    "amountSpendBase": 266.775845845468,
    "averagePrice": 1.1293434813979364,
    "maxDrawdownBase": 301.2815625,
    "order": 60.775312500000005,
    "orderNo": 5,
    "price": 1.07739628378004,
    "profitTarget": 1.1462836336189053,
    "safetyOrderDeviation": 10.202010020000001,
  },
  Object {
    "amountSpendBase": 327.4254550056507,
    "averagePrice": 1.1150496549472588,
    "maxDrawdownBase": 365.095640625,
    "order": 63.81407812500001,
    "orderNo": 6,
    "price": 1.0521762466178404,
    "profitTarget": 1.1317753997714677,
    "safetyOrderDeviation": 12.304030120200002,
  },
]
`;
